snapcore/snapcraft
# meta: implement pass-through of properties to snap.yaml

-----
This PR introduces a new **pass-through** keyword to `snapcraft.yaml` which comes in two forms, a global and an app-local property [as proposed in the forum](https://forum.snapcraft.io/t/support-for-passthrough-into-snap-yaml/4698). As proposed, the key is just defined as a dictionary (object in the schema) and not validated, but instead applied in `snap_packaging.py` where its use will be logged and ambiguous keys will raise an error. The original snapcraft.yaml is used to validate what values have been specified as opposed to default values filled in by Snapcraft.
Note that, as discussed offline, properties like `name` or `command` are not special-cased and not supported by pass-through due to their being strings processed by Snapcraft.

The following tests are added/ modified:

 - tests.unit.test_errors
 - To verify AmbiguousPassThroughError
 - tests.unit.test_meta.PassThroughTestCase
 - To verify that a) a new key not present in the schema can be added b) an existing key can be overridden c) a key both present in the schema and pass-through raising an error d) defaults can be overridden rather than error out because of the check for amiguity e) a warning being logged for using pass-through. Each of these is tested with the global and the apps form.

I locally ran the tests:
 - `./runtests.sh tests/unit` with unrelated failures in [tests.unit.test_lifecycle.CoreSetupTestCase.test_core_setup_if_docker_env](https://bugs.launchpad.net/snapcraft/+bug/1752576), `tests.unit.test_mangling.TestClearExecstack.test_execstack_clears` and `tests.unit.test_elf`.
 - `./runtests.sh tests/integration` with one unrelated failure in [tests.integration.general.test_parser.TestParserWikis](https://bugs.launchpad.net/snapcraft/+bug/1752580)
 - `./runtests.sh static`: Everything passed

Manual test steps:
 - cd tests/integration/snaps/basic
 - Add some pass-through properties:
   pass-through:
     foo: bar
     confinement: next-generation
   apps:
     foo:
       command: echo
       daemon: simple
       pass-through:
         foo: bar
         daemon: complex
 - snapcraft
 - Observe the error "Failed to pass-through keys to snap metadata: 'daemon' is specified as both a regular key and in pass-through. Remove one of the keys.".
 - Fix the error by removing "daemon: simple" from the "foo" app.
 - snapcraft
 - Observe the error "Failed to pass-through keys to snap metadata: 'confinement' is specified as both a regular key and in pass-through. Remove one of the keys.".
 - Fix the error by removing "confinement: classic" from the top.
 - snapcraft
 - Observe that that the build succeeds with two warnings:
   Pass-through is being used to add experimental properties to 'foo' that have not been validated. The snap cannot be released to the store.
   Pass-through is being used to add experimental properties to 'snapcraft.yaml' that have not been validated. The snap cannot be released to the store.
 - grep -C 3 foo prime/meta/snap.yaml
   confinement: next-generation
   grade: stable
   apps:
     foo:
       command: command-foo.wrapper
       daemon: complex
       foo: bar
   foo: bar
 - Observe keys being propagated to snap.yaml without validation.
